Activator panics in rewinder

<!--
/area autoscale
/kind bug
-->

## Expected Behavior

Activator should be resilient to various operating conditions.

## Actual Behavior

Something causes the Activator to panic in `rewinder.go`.

## Steps to Reproduce the Problem

  1. Scale to and from zero a few times.
  2. Look at the restarts / logs for the activator pod.

## Additional Info

Example:
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x4f392c]

goroutine 537 [running]:
io/ioutil.readAll.func1(0xc42016d6d8)
        /home/josephburnett/goroot/src/io/ioutil/ioutil.go:30 +0x104
panic(0x105c140, 0x19ecf60)
        /home/josephburnett/goroot/src/runtime/panic.go:502 +0x229
bytes.(*Buffer).ReadFrom(0xc420174af0, 0x0, 0x0, 0xc42031f940, 0x0, 0x0)
        /home/josephburnett/goroot/src/bytes/buffer.go:205 +0x6c
io/ioutil.readAll(0x0, 0x0, 0x200, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/josephburnett/goroot/src/io/ioutil/ioutil.go:36 +0xb5
io/ioutil.ReadAll(0x0, 0x0, 0x0, 0x0, 0x0, 0x410498, 0x1)
        /home/josephburnett/goroot/src/io/ioutil/ioutil.go:45 +0x3e
github.com/knative/serving/pkg/activator/util.(*rewinder).Read(0xc4205ed2e0, 0xc42041ad38, 0x1, 0x1, 0x1, 0x0, 0x0)
        /home/josephburnett/go/src/github.com/knative/serving/pkg/activator/util/rewinder.go:36 +0xc9
net/http.(*transferWriter).probeRequestBody.func1(0xc420184aa0, 0x7fc2f9c8afa8, 0xc4205ed2e0)
        /home/josephburnett/goroot/src/net/http/transfer.go:207 +0x64
created by net/http.(*transferWriter).probeRequestBody
        /home/josephburnett/goroot/src/net/http/transfer.go:204 +0xa5
```

--- a/pkg/activator/util/rewinder.go
+++ b/pkg/activator/util/rewinder.go
@@ -17,9 +17,11 @@ import (
 	"bytes"
 	"io"
 	"io/ioutil"
+	"sync"
 )
 
 type rewinder struct {
+	sync.Mutex
 	rc io.ReadCloser
 	rs io.ReadSeeker
 }
@@ -30,6 +32,8 @@ func NewRewinder(rc io.ReadCloser) io.ReadCloser {
 }
 
 func (r *rewinder) Read(b []byte) (int, error) {
+	r.Lock()
+	defer r.Unlock()
 	// On the first `Read()`, the contents of `rc` is read into a buffer `rs`.
 	// This buffer is used for all subsequent reads
 	if r.rs == nil {
@@ -46,6 +50,8 @@ func (r *rewinder) Read(b []byte) (int, error) {
 }
 
 func (r *rewinder) Close() error {
+	r.Lock()
+	defer r.Unlock()
 	// Rewind the buffer on `Close()` for the next call to `Read`
 	r.rs.Seek(0, io.SeekStart)
 

--- a/pkg/activator/util/transports.go
+++ b/pkg/activator/util/transports.go
@@ -74,9 +74,13 @@ func (rrt *retryRoundTripper) RoundTrip(r *http.Request) (resp *http.Response, e
 	// The request body cannot be read multiple times for retries.
 	// The workaround is to clone the request body into a byte reader
 	// so the body can be read multiple times.
-	r.Body = NewRewinder(r.Body)
+	if r.Body != nil {
+		rrt.logger.Debugf("Wrapping body in a rewinder.")
+		r.Body = NewRewinder(r.Body)
+	}
 
 	attempts := rrt.retryer.Retry(func() bool {
+		rrt.logger.Debugf("Retrying")
 		resp, err = rrt.transport.RoundTrip(r)
 
 		if err != nil {

